### PR TITLE
feat(cart): add create method to cart driver

### PR DESCRIPTION
**Pull request by griest024**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

No way to create a cart.
Fixes: N/A

## What is the new behavior?
Create a cart with `DaffCartServiceInterface#create`.

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

## Other information

### CODE REVIEW

**Comment by lderrickable** [comment]
I tried running the demo app to test the in memory cart service and it doesn't seem to work. The add to cart request fails.

**Comment by griest024** [comment]
@lderrickable Demo will rely on all the cart interface changes to be reflected in the testing module. I'll add that in a separate PR unless you prefer to have it in this one.

File: libs/cart/testing/src/in-memory-backend/cart.service.integration.spec.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    stubProductImage = new DaffProductImageFactory().create();
+    stubCart = new DaffCartFactory().create({image: stubProductImage});
+
+    TestBed.configureTestingModule({
+      imports: [
+        HttpClientModule,
+        HttpClientInMemoryWebApiModule.forRoot(DaffInMemoryBackendCartService)
+      ],
+      providers: [
+        { provide: DaffCartFactory, useValue: daffCartFactorySpy},
```
Comment (damienwebdev) [inline libs/cart/testing/src/in-memory-backend/cart.service.integration.spec.ts]: Whats the purpose of these spies?

File: libs/cart/testing/src/drivers/testing/cart.service.ts [hunk 2]
```diff
@@ -24,4 +24,8 @@ export class DaffTestingCartService implements DaffCartServiceInterface<DaffCart
   clear(): Observable<void> {
     return of();
   }
+
+  create() {
+    return this.get();
```
Comment (damienwebdev) [inline libs/cart/testing/src/drivers/testing/cart.service.ts]: Let's just return a:

```
{
 id: this.cartFactory.create().id
}
```

as that will likely be the most common case.

File: libs/cart/src/drivers/interfaces/cart-service.interface.ts [hunk 3]
```diff
@@ -11,7 +10,12 @@ export interface DaffCartServiceInterface<T extends DaffCart> {
 	/**
 	 * Retrieve a cart
 	 */
-	get(id: T['id']): Observable<T>;
+  get(id: T['id']): Observable<T>;
+
+  /**
+   * Creates a cart.
+   */
+  create(): Observable<Partial<T>>
```
Comment (damienwebdev) [inline libs/cart/src/drivers/interfaces/cart-service.interface.ts]: I think we should probably pull back this interface to something more minimal. 

```
{
 id: T['id']
}
```

as that is likely the common case.

File: libs/cart/testing/src/in-memory-backend/cart.service.unit.spec.ts [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+  });
+
+  describe('processing a clear request', () => {
+    let reqInfoStub;
+    let result;
+
+    beforeEach(() => {
+      cartTestingService.carts.push(stubCart);
+      //add cartItems
+      cartTestingService.post({
```
Comment (damienwebdev) [inline libs/cart/testing/src/in-memory-backend/cart.service.unit.spec.ts]: Can you just stub a cart with items in it instead?

File: libs/cart/testing/src/drivers/in-memory/cart.service.ts [hunk 5]
```diff
@@ -12,15 +12,19 @@ export class DaffInMemoryCartService implements DaffCartServiceInterface<DaffCar
...
   }
 
-  clear(): Observable<void> {
-    return this.http.post<void>(this.url + '/clear', {});
+  clear(cartId: string): Observable<void> {
+    return this.http.post<void>(this.url + '/clear', {cartId});
+  }
+
+  create(): Observable<Partial<DaffCart>> {
+    return this.http.post<Partial<DaffCart>>(this.url + '/create', {});
```
Comment (damienwebdev) [inline libs/cart/testing/src/drivers/in-memory/cart.service.ts]: `create` should just be a post to `this.url`

File: libs/cart/testing/src/in-memory-backend/cart.service.integration.spec.ts (hunk 1 above)
Comment (griest024) [reply libs/cart/testing/src/in-memory-backend/cart.service.integration.spec.ts]: They don't seem to have a purpose actually. They were in the spec from before and I didn't touch them. They should be removed.

File: libs/cart/testing/src/drivers/testing/cart.service.ts (hunk 2 above)
Comment (griest024) [reply libs/cart/testing/src/drivers/testing/cart.service.ts]: Changed in recent push.

File: libs/cart/src/drivers/interfaces/cart-service.interface.ts (hunk 3 above)
Comment (griest024) [reply libs/cart/src/drivers/interfaces/cart-service.interface.ts]: Changed in recent push.

File: libs/cart/testing/src/in-memory-backend/cart.service.unit.spec.ts (hunk 4 above)
Comment (griest024) [reply libs/cart/testing/src/in-memory-backend/cart.service.unit.spec.ts]: Changed in recent push.

File: libs/cart/testing/src/drivers/in-memory/cart.service.ts (hunk 5 above)
Comment (griest024) [reply libs/cart/testing/src/drivers/in-memory/cart.service.ts]: Changed in recent push.

File: libs/cart/testing/src/in-memory-backend/cart.service.integration.spec.ts (hunk 1 above)
Comment (griest024) [reply libs/cart/testing/src/in-memory-backend/cart.service.integration.spec.ts]: Changed in recent push.

**Comment by griest024** [comment]
[e12e0f3](https://github.com/graycoreio/daffodil/pull/643/commits/e12e0f3e93b3554e147843f1094e4415030ad34a) also includes some changes in the main package necessary for this. I will separate that into another PR if its out of the scoppe of this PR.

File: libs/cart/src/models/cart-item-input.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
 export interface DaffCartItemInput {
-  productId: string;
+  productId: number;
```
Comment (damienwebdev) [inline libs/cart/src/models/cart-item-input.ts]: Should be off the Type of DaffProduct.

File: libs/cart/testing/src/drivers/in-memory/cart.service.spec.ts [hunk 7]
```diff
@@ -57,46 +61,59 @@ describe('Driver | In Memory | Cart | CartService', () => {
...
-        req.flush(mockCart);
+        req.flush(null);
       });
     });
   });
+
+  describe('create | creating a cart', () => {
+    it('should send a post request to `api/cart/create` and return the cart', done => {
+      cartService.create().subscribe(result => {
+        expect(result).toEqual(mockCart);
```
Comment (damienwebdev) [inline libs/cart/testing/src/drivers/in-memory/cart.service.spec.ts]: This expectation should be failing based upon the interface described.

File: libs/cart/src/models/cart-item-input.ts (hunk 6 above)
Comment (griest024) [reply libs/cart/src/models/cart-item-input.ts]: Changed this as well as `product_id` in `DaffCartItem`.

File: libs/cart/testing/src/drivers/in-memory/cart.service.spec.ts (hunk 7 above)
Comment (griest024) [reply libs/cart/testing/src/drivers/in-memory/cart.service.spec.ts]: Fixed the message.

Comment (damienwebdev) [reply libs/cart/testing/src/drivers/in-memory/cart.service.spec.ts]: No, I think that the test should actually be failing as described. `create` does not return a `mockcart`.

Comment (griest024) [reply libs/cart/testing/src/drivers/in-memory/cart.service.spec.ts]: Changed the expected and flushed value to just an object with ID.

**Pull request merged**